File: backend/internal/services/message_service.go
// ============================================================
// internal/services/message_service.go
// ✅ [ANGGOTA 1] SendFromDosen — dosen kirim pesan kendala
// ✅ [ANGGOTA 2] GetAdminMessages, MarkAsRead — admin baca pesan
// ============================================================

package services

import (
	"gorm.io/gorm"
	"sistem-presensi-dosen/internal/models"
)

type MessageService struct {
	db *gorm.DB
}

func NewMessageService(db *gorm.DB) *MessageService {
	return &MessageService{db: db}
}

// ─────────────────────────────────────────────────────────────
// [ANGGOTA 1] SendFromDosen
// ─────────────────────────────────────────────────────────────
// Dosen mengirim laporan kendala ke admin.
// sessionID adalah *uint (pointer) karena bisa null — dosen bisa
// melapor kendala umum yang tidak terkait sesi tertentu.
func (s *MessageService) SendFromDosen(dosenID uint, judul, isi string, sessionID *uint) (*models.Message, error) {
	msg := models.Message{
		DosenID:   dosenID,
		SessionID: sessionID, // null jika tidak ada sesi terkait
		Judul:     judul,
		Isi:       isi,
		Status:    "unread", // semua pesan baru selalu masuk sebagai 'unread'
	}

	if err := s.db.Create(&msg).Error; err != nil {
		return nil, err
	}

	// Preload Dosen agar response menyertakan nama dosen (bukan hanya ID)
	if err := s.db.Preload("Dosen").First(&msg, msg.ID).Error; err != nil {
		return nil, err
	}

	return &msg, nil
}

// ─────────────────────────────────────────────────────────────
// [ANGGOTA 2] GetAdminMessages
// ─────────────────────────────────────────────────────────────
// Ambil semua pesan dari dosen dengan filter status opsional.
// status: "" = semua | "unread" | "read"
func (s *MessageService) GetAdminMessages(status string) ([]models.Message, error) {
	var messages []models.Message

	query := s.db.
		Preload("Dosen").
		Preload("Session.Schedule").
		Order("created_at DESC")

	if status == "unread" || status == "read" {
		query = query.Where("status = ?", status)
	}

	err := query.Find(&messages).Error
	return messages, err
}

// ─────────────────────────────────────────────────────────────
// [ANGGOTA 1] GetDosenMessages
// ─────────────────────────────────────────────────────────────
// Ambil semua pesan yang pernah dikirim oleh dosen tertentu.
func (s *MessageService) GetDosenMessages(dosenID uint) ([]models.Message, error) {
	messages := make([]models.Message, 0) // make agar JSON encode [] bukan null
	err := s.db.
		Preload("Session.Schedule").
		Where("dosen_id = ?", dosenID).
		Order("created_at DESC").
		Find(&messages).Error
	return messages, err
}

// ─────────────────────────────────────────────────────────────
// [ANGGOTA 2] MarkAsRead
// ─────────────────────────────────────────────────────────────
// Tandai pesan sebagai sudah dibaca oleh admin.
func (s *MessageService) MarkAsRead(messageID uint) (*models.Message, error) {
	var msg models.Message

	if err := s.db.First(&msg, messageID).Error; err != nil {
		return nil, err
	}

	if err := s.db.Model(&msg).Update("status", "read").Error; err != nil {
		return nil, err
	}

	return &msg, nil
}

File: frontend/src/services/messageService.js
// ============================================================
// services/messageService.js
// ✅ [ANGGOTA 1 - KAMU] Kirim pesan kendala ke admin.
// ✅ [ANGGOTA 2]         Ambil dan tandai pesan dari dosen.
// ============================================================

import api from './api'

// ─── Dosen: Kirim pesan kendala ke admin ─────────────────────
// TODO: POST /api/messages/send
// Body: { judul, isi, session_id (opsional) }
// Gunakan ketika dosen tidak bisa record kehadiran karena kendala
// (GPS mati, session expired, dll)
// Response: { id, status: 'sent', created_at }
export const sendIssueToAdmin = async ({ judul, isi, sessionId }) => {
  const response = await api.post('/messages/send', {
    judul,
    isi,
    session_id: sessionId || null,
  })
  return response.data
}

// ─── Admin: Ambil semua pesan masuk dari dosen ───────────────
// TODO: GET /api/admin/messages?status=unread
// Response: array of { id, dosen_nama, judul, isi, session_id, status, created_at }
export const getAdminMessages = async (status = '') => {
  const response = await api.get('/admin/messages', { params: { status } })
  return response.data
}

// ─── Admin: Tandai pesan sebagai sudah dibaca ────────────────
// TODO: PATCH /api/admin/messages/:id/read
// Response: { id, status: 'read' }
export const markMessageAsRead = async (messageId) => {
  const response = await api.patch(`/admin/messages/${messageId}/read`)
  return response.data
}
